grpc/src/abci_proofs.rs

```diff
@@ -0,0 +1,227 @@
+    type Error = ProofError;
+
+    fn try_from(value: ProofOp) -> Result<Self, Self::Error> {
+        let spec = match value.r#type.as_str() {
+            "ics23:iavl" => ics23::iavl_spec(),
+            "ics23:simple" => ics23::tendermint_spec(),
+            _ => return Err(ProofError::UnsupportedSpec(value.r#type.clone())),
+        };
+        Ok(Self {
+            key: value.key,
+            spec,
+            proof: CommitmentProof::decode(value.data.as_slice())?,
+        })
+    }
+}
+
+impl CommitmentOp {
+    /// A helper to extract the underlying existance proof (if any) for a given key.
+    fn get_existence_proof(&self, key: impl AsRef<[u8]>) -> Option<&ExistenceProof> {
+        match self.proof.proof.as_ref()? {
+            Proof::Exist(proof) => Some(proof),
+            Proof::Batch(batch) => batch.entries.iter().find_map(|entry| {
+                if let ics23::batch_entry::Proof::Exist(proof) = entry.proof.as_ref()? {
+                    if key.as_ref() == proof.key {
+                        Some(proof)
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
+            }),
+            _ => None,
+        }
+    }
+}
+
+/// A chain of proofs that is used to verify a leaf value in multiple nested merkle trees.
+pub struct ProofChain(Vec<CommitmentOp>);
+
+impl ProofChain {
+    /// Verifies that a leaf value exists in nested merkle trees.
+    ///
+    /// Root is an expected root hash of the uppermost merkle tree.
+    /// A leaf is a value we want to prove inclusion of, starting from the lowermost merkle tree.
+    /// Each proof corresponds to one tree in a chain, and proves the inclusion of its leaf up to a leaf of the next proof,
+    /// until the last proof when the value is proven up to the expected uppermost root. Provided keys must match the leaves'
+    /// keys starting from the lowermost leaf.
+    ///
+    /// ```text
+    ///                         root
+    ///                       /      \
+    ///                      /        \
+    ///                     o          o
+    ///                   /   \      /   \
+    ///                  /     \    /     \
+    ///             inner leaf  o  o       o
+    ///           (with 2nd key)
+    ///                 ==
+    ///             inner root
+    ///              /      \
+    ///             /        \
+    ///            o          o
+    ///          /   \      /   \
+    ///         /     \    /     \
+    ///       leaf     o  o       o
+    ///  (with 1st key)
+    /// ```
+    pub fn verify_membership(
+        &self,
+        root: impl AsRef<[u8]>,
+        keys: impl IntoIterator<Item = impl AsRef<[u8]>>,
+        leaf: impl AsRef<[u8]>,
+    ) -> Result<(), ProofError> {
+        let root = root.as_ref();
+        let mut current_leaf = leaf.as_ref();
+
+        for (i, key) in keys.into_iter().enumerate() {
```

Comment:
This need to check if length of proofs and keys is equal, otherwise allows for proof crafting